File: libs/power-log-parser/src/lib/parsers/game-end-parser.ts
import { GameTag } from '@firestone-hs/reference-data';
import { ActionParser } from '../action-parser';
import { GameEventProvider } from '../game-event';
import { Node, TagChange } from '../models';
import { Logger } from '../logger';
import { xmlFromReplay } from '../replay-converter';
import { GameState } from '../state/game-state';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

const STATE_COMPLETE = 3;

export class GameEndParser implements ActionParser {
	readonly ParserName = 'GameEndParser';

	private GameState: GameState;
	private ParserState: ParserState;
	private StateFacade: StateFacade;

	constructor(parserState: ParserState, stateFacade: StateFacade) {
		this.ParserState = parserState;
		this.GameState = parserState.GameState;
		this.StateFacade = stateFacade;
	}

	AppliesOnNewNode(node: Node, stateType: StateType): boolean {
		let tagChange: TagChange | null = null;
		return (
			stateType === StateType.PowerTaskList &&
			node.Type === TagChange &&
			(((tagChange = node.Object as TagChange).Name === (GameTag.STATE as number) &&
				tagChange.Value === STATE_COMPLETE) ||
				(this.ParserState.IsBattlegrounds() &&
					tagChange!.Name === (GameTag.TAG_PLAYER_CONCEDED_OR_DISCONNECTED as number) &&
					tagChange!.Value === 1))
		);
	}

	AppliesOnCloseNode(_node: Node, _stateType: StateType): boolean {
		return false;
	}

	CreateGameEventProviderFromNew(node: Node): GameEventProvider[] | null {
		Logger.Log('Parsing end game', node.CreationLogLine);
		const tagChange = node.Object as TagChange;
		if (tagChange.Name === (GameTag.TAG_PLAYER_CONCEDED_OR_DISCONNECTED as number)) {
			const isPlayer = tagChange.Entity === this.StateFacade.LocalPlayer?.Id;
			if (!isPlayer) {
				return null;
			}
		}

		for (const player of this.ParserState.getPlayers()) {
			const gsPlayer = this.StateFacade.GetPlayers().find((p) => p.Id === player.Id);
			player.Name = gsPlayer?.Name ?? player.Name;
		}

		const gameStateReport = this.GameState.BuildGameStateReport(this.StateFacade);
		Logger.Log('gameStateReport built', '');

		let replayXml: string | null = null;
		try {
			Logger.Log('Will convert to xml', '');
			replayXml = xmlFromReplay(this.StateFacade.GSReplay);
			Logger.Log('XML converted', '');
		} catch (ex) {
			Logger.Log('Could not convert replay to xml', `${ex}`);
		}

		Logger.Log('Enqueuing GAME_END event', '');
		this.ParserState.EndCurrentGame();
		return [
			GameEventProvider.Create(
				tagChange.TimeStamp,
				'GAME_END',
				() => ({
					Type: 'GAME_END',
					Value: {
						LocalPlayer: this.StateFacade.LocalPlayer,
						OpponentPlayer: this.StateFacade.OpponentPlayer,
						GameStateReport: gameStateReport,
						ReplayXml: replayXml,
						FormatType: this.ParserState.CurrentGame.FormatType,
						GameType: this.ParserState.CurrentGame.GameType,
						ScenarioID: this.ParserState.CurrentGame.ScenarioID,
						Spectating: this.StateFacade.Spectating,
					},
				}),
				true,
				node,
			),
		];
	}

	CreateGameEventProviderFromClose(_node: Node): GameEventProvider[] | null {
		return null;
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/GameEndParser.cs
using HearthstoneReplays.Parser;
using HearthstoneReplays.Parser.ReplayData;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using System;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using System.Collections.Generic;

namespace HearthstoneReplays.Events.Parsers
{
    public class GameEndParser : ActionParser
    {
        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade StateFacade { get; set; }

        public GameEndParser(ParserState ParserState, StateFacade helper)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.StateFacade = helper;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            TagChange tagChange = null;
            return stateType == StateType.PowerTaskList
                && node.Type == typeof(TagChange)
                && (
                    ((tagChange = node.Object as TagChange).Name == (int)GameTag.STATE && tagChange.Value == (int)State.COMPLETE)
                    || (ParserState.IsBattlegrounds() && tagChange.Name == (int)GameTag.TAG_PLAYER_CONCEDED_OR_DISCONNECTED && tagChange.Value == 1)
                );
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            return false;
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            Logger.Log("Parsing end game", node.CreationLogLine);
            var tagChange = node.Object as TagChange;
            if (tagChange.Name == (int)GameTag.TAG_PLAYER_CONCEDED_OR_DISCONNECTED)
            {
                var isPlayer = tagChange.Entity == StateFacade.LocalPlayer.Id;
                if (!isPlayer)
                {
                    return null;
                }
            }
            var replayCopy = StateFacade.GSReplay;
            // Update the name info
            foreach (var player in ParserState.getPlayers())
            {
                var gsPlayer = StateFacade.GetPlayers().Find(p => p.Id == player.Id);
                player.Name = gsPlayer?.Name ?? player.Name;
            }
            Logger.Log("Will convert to xml", "");
            var xmlReplay = new ReplayConverter().xmlFromReplay(replayCopy);
            Logger.Log("XML converted", "");
            var gameStateReport = GameState.BuildGameStateReport(StateFacade);
            Logger.Log("gameStateReport built", "");
            //var gameState = GameEvent.BuildGameState(ParserState, StateFacade, GameState, tagChange, null);
            Logger.Log("Enqueuing GAME_END event", "");
            ParserState.EndCurrentGame();
            return new List<GameEventProvider> { GameEventProvider.Create(
                tagChange.TimeStamp,
                "GAME_END",
                () => new GameEvent
                {
                    Type = "GAME_END",
                    Value = new
                    {
                        LocalPlayer = StateFacade.LocalPlayer,
                        OpponentPlayer = StateFacade.OpponentPlayer,
                        GameStateReport = gameStateReport,
                        FormatType = ParserState.CurrentGame.FormatType,
                        GameType = ParserState.CurrentGame.GameType,
                        ScenarioID = ParserState.CurrentGame.ScenarioID,
                        //Game = ParserState.CurrentGame,
                        ReplayXml = xmlReplay,
                        Spectating = StateFacade.Spectating,
                    }
                },
                true,
                node) };
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            return null;
        }
    }
}
